cmd/dp-import-api/main.go

```diff
@@ -11,147 +12,125 @@ import (
 
-	_ = api.CreateImportAPI(router, mongoDataStore, jobService, auditor)
+	api.CreateImportAPI(cfg.Host, cfg.BindAddr, cfg.ZebedeeURL, mongoDataStore, jobService, auditor)
 
-	// signals the web server shutdown, so a graceful exit is required
-	httpErrChannel := make(chan error)
-	// launch web server in background
 	go func() {
-		log.Debug("listening...", log.Data{"bind_address": cfg.BindAddr})
-		if err := httpServer.ListenAndServe(); err != nil {
-			log.Error(err, nil)
-			httpErrChannel <- err
-			return
+		select {
+		case err := <-dataBakerProducer.Errors():
+			log.ErrorC("kafka databaker producer", err, nil)
+		case err := <-directProducer.Errors():
+			log.ErrorC("kafka direct producer", err, nil)
+		case err := <-auditProducer.Errors():
+			log.ErrorC("kafka audit producer", err, nil)
 		}
-		httpErrChannel <- errors.New("http server completed - with no error")
 	}()
 
-	shutdownGracefully := func(httpDead bool) {
-		ctx, cancel := context.WithTimeout(context.Background(), cfg.GracefulShutdownTimeout)
+	// block until a fatal error occurs
+	select {
+	case sig := <-signals:
+		log.Error(errors.New("os signal received"), log.Data{"signal": sig.String()})
+	}
+
+	log.Info(fmt.Sprintf("Shutdown with timeout: %s", cfg.GracefulShutdownTimeout), nil)
+	ctx, cancel := context.WithTimeout(context.Background(), cfg.GracefulShutdownTimeout)
+
+	// Gracefully shutdown the application closing any open resources.
+	go func() {
 		defer cancel()
 
-		if !httpDead {
-			if err = httpServer.Shutdown(ctx); err != nil {
-				log.Error(err, nil)
-			}
+		log.Info("closing http server for healthcheck", nil)
+		if err = api.Close(ctx); err != nil {
+			logIfError(err, "unable to close api server")
 		}
 
-		if err = auditProducer.Close(ctx); err != nil {
-			log.Error(err, nil)
+		if serviceList.MongoDataStore {
+			log.Info("closing mongo data store", nil)
+			// mongo.Close() may use all remaining time in the context
+			logIfError(mongoDataStore.Close(ctx), "unable to close mongo data store")
 		}
 
-		if err = dataBakerProducer.Close(ctx); err != nil {
-			log.Error(err, nil)
+		if serviceList.DataBakerProducer {
+			log.Info("closing data baker producer", nil)
+			logIfError(dataBakerProducer.Close(ctx), "unable to close data baker producer")
 		}
 
-		if err = directProducer.Close(ctx); err != nil {
-			log.Error(err, nil)
+		if serviceList.DirectProducer {
+			log.Info("closing direct producer", nil)
+			logIfError(directProducer.Close(ctx), "unable to close direct producer")
 		}
 
-		// mongo.Close() may use all remaining time in the context - do this last!
-		if err = mongoDataStore.Close(ctx); err != nil {
-			log.Error(err, nil)
+		// Close audit producer last, to maximise capturing all events
+		if serviceList.AuditProducer {
+			log.Info("closing audit producer", nil)
+			logIfError(auditProducer.Close(ctx), "unable to close audit producer")
 		}
+	}()
+
+	// wait for shutdown success (via cancel) or failure (timeout)
+	<-ctx.Done()
```

Comment:
Im a bit rusty on go context but is there potential for this to block indefinitely?

Reply:
No as there is a timeout (`GracefulShutdownTimeout` environment config) set on the context which will send a message to ctx.Done() channel after said time and hence force a shutdown.

If context.WithTimeout was not set you would be correct as the graceful shutdown may fail to ever call the cancel function (which also sends a message to ctx.Done( ) channel).

Reply:
👍